### PR TITLE
feat: add dns discovery

### LINKED ISSUE #3517: [DISCUSS] add per node DNS selector
When resolving domain, currently we take one of the DNS record randomly. When we want to do service discovery via DNS, sometimes we hope the domain resolving will return all the records. For example, with:

```
"nodes": {
    "test.consul.service": 1
}
```

We may want to get all IPs instead of a random one.

So my purpose is to provide a per-node dns selector to control the behavior.

For example,

```
"nodes": {
    {"host": "test.consul.service", "weight": 1, "selector": "random"}
}
```

is equal to previous configuration.

While

```
"nodes": {
    {"host": "test.consul.service", "weight": 1, "selector": "all"},
    {"host": "1.1.1.1", "weight": 1}
}
```

with `test.consul.service` be resolved as `1.1.1.2` and `1.1.1.3` will get this result:

```
"nodes": {
    {"host": "1.1.1.1", "weight": 2},
    {"host": "1.1.1.2", "weight": 1},
    {"host": "1.1.1.3", "weight": 1},
}
```

Note that all the IPs from `test.consul.service` share the same weight.
And the selector doesn't affect SRV record, which is a difference story.

### PR DESCRIPTION
**Pull request by spacewander**
Fix #3517
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Review by tokers — APPROVED** [review]

File: apisix/upstream.lua [hunk 1]
```diff
@@ -179,7 +179,12 @@ function _M.set_by_route(route, api_ctx)
         if not dis then
             return 500, "discovery " .. up_conf.discovery_type .. " is uninitialized"
         end
-        local new_nodes = dis.nodes(up_conf.service_name)
+
+        local new_nodes, err = dis.nodes(up_conf.service_name)
+        if not new_nodes then
+            return http_code_upstream_unavailable, "no valid upstream node: " .. (err or "nil")
```
Comment (membphis) [inline apisix/upstream.lua:185]: `http_code_upstream_unavailable`:

For static variables, use capital letters

**Review by membphis — APPROVED** [review]
LGTM, except minor style things

File: apisix/upstream.lua (hunk 1 above)
Comment (spacewander) [reply apisix/upstream.lua:185]: @membphis 
I have submitted another PR to use the number literal directly: https://github.com/apache/apisix/pull/3672, which is more consistent in the style.

Comment (spacewander) [reply apisix/upstream.lua:185]: I changed my idea and followed your suggestion.

**Pull request merged**